pkg/collections.go

```diff
@@ -0,0 +1,264 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+// Author Gergely Brautigam
+//
+package pkg
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/arangodb/go-driver"
+	"github.com/briandowns/spinner"
+	"github.com/cenkalti/backoff"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+)
+
+// copyCollections copies all collections for a database.
+func (c *copier) copyCollections(ctx context.Context, db driver.Database) error {
+	log := c.Logger
+	var (
+		sourceDB      driver.Database
+		list          []driver.Collection
+		destinationDB driver.Database
+	)
+	c.backoffCall(ctx, func() error {
+		sdb, err := c.sourceClient.Database(ctx, db.Name())
+		if err != nil {
+			return err
+		}
+		sourceDB = sdb
+		colls, err := sourceDB.Collections(ctx)
+		if err != nil {
+			c.Logger.Error().Err(err).Msg("Failed to list collections for source database.")
+			return err
+		}
+		list = colls
+		ddb, err := c.destinationClient.Database(ctx, sourceDB.Name())
+		if err != nil {
+			c.Logger.Error().Err(err).Msg("Failed to get destination database.")
+			return nil
+		}
+		destinationDB = ddb
+		return nil
+	})
+
+	collections := c.filterCollections(list)
+	readCtx := driver.WithQueryStream(ctx, true)
+	readCtx = driver.WithQueryBatchSize(readCtx, c.BatchSize)
+	restoreCtx := driver.WithIsRestore(ctx, true)
+	var g errgroup.Group
+	sem := semaphore.NewWeighted(int64(c.Parallel))
+	s := spinner.New(spinner.CharSets[34], 100*time.Millisecond)
+	s.Start()
+	for _, sourceColl := range collections {
+		sourceColl := sourceColl
+		g.Go(func() error {
+			if err := sem.Acquire(ctx, 1); err != nil {
+				return err
+			}
+			defer sem.Release(1)
+			var props driver.CollectionProperties
+			c.backoffCall(ctx, func() error {
+				sourceProps, err := sourceColl.Properties(ctx)
+				if err != nil {
+					c.Logger.Error().Err(err).Msg("Failed to get properties.")
+					return err
+				}
+				props = sourceProps
+				return nil
+			})
+			if props.IsSystem {
+				// skip system collections
+				return nil
+			}
+			var destinationColl driver.Collection
+			c.backoffCall(ctx, func() error {
+				dColl, err := c.ensureDestinationCollection(ctx, destinationDB, sourceColl, props)
+				if err != nil {
+					c.Logger.Error().Err(err).Msg("Failed to ensure destination collection.")
+					return err
+				}
+				destinationColl = dColl
+				return nil
+			})
+			bindVars := map[string]interface{}{
+				"@c": sourceColl.Name(),
+			}
+			var cursor driver.Cursor
+			c.backoffCall(ctx, func() error {
+				cr, err := sourceDB.Query(readCtx, "FOR d IN @@c RETURN d", bindVars)
+				if err != nil {
+					c.Logger.Error().Err(err).Str("collection", sourceColl.Name()).Msg("Failed to query source database for collection.")
+					return err
+				}
+				cursor = cr
+				return nil
+			})
+			batch := make([]interface{}, 0, c.BatchSize)
```

Comment:
defer cursor.Close()

Reply:
Yeah this was a remnant of running in a for loop instead of concurrently. The loop prevented a defer close.